Implement content type negotiation

mpalmer opened the pull request:
This is really what I (and I think most people) need from an Accept header implementation.  So now there is one!

I used a slice of `MediaType`s rather than `MediaTypeBuf` because `MediaType` is const-constructable, and the way I define my set of acceptable media types is via module-level `const`.

maxcountryman commented:
Looks like the slice is being compared to the vec and that's probably not what was intended.

maxcountryman commented:
I would suggest making `negotiate` a bit more flexible:

```diff
diff --git a/src/lib.rs b/src/lib.rs
index 8df0c5e..a6c074d 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -100,11 +100,14 @@ impl Accept {
     ///
     /// Don't put wildcard types or the `q` parameter in the `available` list;
     /// if you do, all bets are off as to what might happen.
-    pub fn negotiate<'a, 's>(&'s self, available: &[&'a MediaType<'a>]) -> Option<&MediaType<'a>> {
+    pub fn negotiate<'a, I>(&self, available: I) -> Option<&MediaType<'a>>
+    where
+        I: IntoIterator<Item = &'a MediaType<'a>>,
+    {
         let mut best = (0.0, (0, 0), None);
 
         available
-            .iter()
+            .into_iter()
             .enumerate()
             .map(|(pt, at)| {
                 if let Some(matched_range) = self
@@ -116,7 +119,7 @@ impl Accept {
                     (
                         Self::parse_q_param(matched_range.1),
                         (matched_range.0, pt),
-                        Some(at)
+                        Some(at),
                     )
                 } else {
                     (0.0, (0, 0), None)
@@ -128,7 +131,7 @@ impl Accept {
                 }
             });
 
-        best.2.copied()
+        best.2
     }
 
     fn parse(mut s: &str) -> Result<Self, HeaderError> {
```

maxcountryman commented on src/lib.rs:
```diff
@@ -56,6 +57,80 @@ impl Accept {
+    /// Don't put wildcard types or the `q` parameter in the `available` list;
+    /// if you do, all bets are off as to what might happen.
+    pub fn negotiate<'a, 's>(&'s self, available: &[&'a MediaType<'a>]) -> Option<&MediaType<'a>> {
+        let mut best = (0.0, (0, 0), None);
```
This might be worth pulling out into its own type, `BestMediaType`, for readability.

mpalmer replied:
Do you mean that to negotiate an acceptable content type, you'd do something like this:

```rust
use accept_headers::{Accept, BestMediaType};

// ...

let chosen_media_type = BestMediaType::new(Accept::from_str("text/plain, text/html")?).negotiate(...)?;
```

Or did you have something else in mind?

maxcountryman replied:
No just a private struct to help with readability.

mpalmer replied:
Oooooooohkay, based on #4 now I understand what you were getting at.  Definitely much better than my tuple madness.

maxcountryman commented on src/lib.rs:
```diff
@@ -184,6 +259,31 @@ const fn is_ows(c: char) -> bool {
+struct MediaRange<'a>(&'a MediaTypeBuf);
+
+impl MediaRange<'_> {
+    fn matches(&self, other: &MediaType) -> bool {
```
This looks a lot like `eq`. Should we impl `PartialEq` here?

mpalmer replied:
I hadn't considered it that way, but... I guess so.  The only potential downside I can see is that we'd be able to do `media_range == media_type`, but I don't see how to do `media_type == media_range` -- which isn't a problem, per se, but could perhaps confuse someone in the future?  For an internal-only type, though, it doesn't seem like a big deal.

mpalmer commented:
> I would suggest making negotiate a bit more flexible

Aha!  I tried a few different approaches to make the method more general, but couldn't figure out how.  Glad you're on the case.  I've updated my PR to suit, which also seemed to fix up all the Clippy complaints.  I also switched to the `PartialEq` impl you suggested.  With clarity on the final point, `BestMediaType`, I can sort that out and we should be ready to go.

maxcountryman commented:
We can forgo the BestMediaType since that's just cosmetic and internal.

Thanks for putting this together!

maxcountryman commented:
These changes are incorporated into the `0.1.2` release.

I also did [some clean up](https://github.com/maxcountryman/headers-accept/pull/4/files) that added the `BestMediaType` and refactored the media range equality logic.
